### PR TITLE
Fix autotuner crash when input tensor is None

### LINKED ISSUE #2749: [Bug] Autotuning fails with trtllm_fp8_block_scale_routed_moe
routing_logits is None for trtllm_fp8_block_scale_routed_moe which causes this error when the inputs are passed to the autotuner. Using flashinfer 0.6.4

```
  File "/sgl-workspace/sglang/python/sglang/srt/layers/moe/moe_runner/flashinfer_trtllm.py", line 313, in fused_experts_none_to_flashinfer_trtllm_fp8
    output = trtllm_fp8_block_scale_routed_moe(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/fused_moe/core.py", line 2478, in trtllm_fp8_block_scale_routed_moe
    return get_trtllm_moe_sm100_module().trtllm_fp8_block_scale_moe(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/fused_moe/core.py", line 1683, in trtllm_fp8_block_scale_moe_op
    _, tactic = tuner.choose_one(
                ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/autotuner.py", line 470, in choose_one
    tensors = self._prepare_input_tensors(p, inputs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/autotuner.py", line 792, in _prepare_input_tensors
    tensor = self._create_tensor_like(
             ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/autotuner.py", line 771, in _create_tensor_like
    dtype = origin_tensor.dtype
            ^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'dtype'
```

### PR DESCRIPTION
**Pull request by he-yufeng**
Fixes #2749.

`trtllm_fp8_block_scale_routed_moe` passes `routing_logits=None` for non-routed calls, but `_prepare_input_tensors` assumes all inputs are tensors and crashes in `_create_tensor_like` trying to access `.dtype` on `None`.

Fix: skip `None` inputs and preserve them as-is. This matches the existing pattern in `_prepare_input_tensors_with_batches` which already handles non-tensor inputs with `isinstance(t, torch.Tensor)` checks.

## Summary by CodeRabbit

* **Bug Fixes**
  * Preserve missing inputs during input preparation so None entries are retained and not treated as tensors, preventing errors when some inputs are absent.
  * Relax and align routing and token-count validations to allow empty routing data when appropriate and ensure checks use actual token counts for consistency.

* **Tests**
  * Add regression tests verifying None input handling and safe fallback behavior when routing/tuning data is absent.

### CODE REVIEW

**Comment by trevor-m** [comment]
Hi @he-yufeng I tried your PR, but now I get this error:
```
  File "/sgl-workspace/sglang/python/sglang/srt/layers/moe/moe_runner/flashinfer_trtllm.py", line 333, in fused_experts_none_to_flashinfer_trtllm_fp8
    output = trtllm_fp8_block_scale_routed_moe(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/fused_moe/core.py", line 2478, in trtllm_fp8_block_scale_routed_moe
    return get_trtllm_moe_sm100_module().trtllm_fp8_block_scale_moe(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/fused_moe/core.py", line 1683, in trtllm_fp8_block_scale_moe_op
    _, tactic = tuner.choose_one(
                ^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/autotuner.py", line 480, in choose_one
    valid_tactics = r.get_valid_tactics(tensors, p)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/dist-packages/flashinfer/fused_moe/core.py", line 1043, in get_valid_tactics
    num_tokens = routing_logits.shape[0]
                 ^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'shape'
```

**Comment by he-yufeng** [comment]
Thanks for testing @trevor-m! The crash in `get_valid_tactics` was caused by the same root issue — `MoERunner.get_valid_tactics()` and `MoERunner.forward()` both used `routing_logits.shape[0]` to get `num_tokens`, which fails when `routing_logits` is `None`.

Fixed in 9043934:
- Changed both methods to use `hidden_states.shape[0]` (always available)
- Guarded the shape assertions for `topk_ids`/`expert_weights` that can be empty(0) tensors in pre-computed routing mode

Could you try again with the latest commit?

**Comment by trevor-m** [comment]
@he-yufeng Thanks, it's working now

File: flashinfer/fused_moe/core.py [hunk 1]
```diff
@@ -1093,12 +1093,20 @@ def forward(
...
-            assert expert_weights.shape[0] == num_tokens, (
-                "expert_weights's first dimension must be batch size."
-            )
+            if routing_logits is not None:
+                assert routing_logits.shape[0] == num_tokens, (
+                    "routing_logits's first dimension must be batch size."
+                )
+            # topk_ids/expert_weights can be empty(0) when routing_logits is provided,
+            # or real tensors when pre-computed routing is used.
+            if topk_ids.numel() > 0:
```
Comment (samuellees) [inline flashinfer/fused_moe/core.py]: `expert_weights` is checked like `if expert_weights is not None and expert_weights.numel() > 0:`

Could you keep a similar check style for `topk_ids`, please?

**Comment by he-yufeng** [comment]
Good catch, updated topk_ids check to match the expert_weights style.

**Comment by samuellees** [comment]
> Good catch, updated topk_ids check to match the expert_weights style.

Thanks @he-yufeng ! Could you add a smoke test for your code path, please? I believe the PR will be moved forward very fast once the test is ready ^ ^

**Comment by he-yufeng** [comment]
Added two smoke tests in `test_autotuner_core.py` — one for `_prepare_input_tensors` and one for `choose_one`, both with a `None` optional tensor. Thanks for the nudge!

**Comment by samuellees** [comment]
Hi @he-yufeng , the CI seems passed. Some error are un-relative with this PR. 
But could you take a look at the pre-commit check fail? 
https://github.com/flashinfer-ai/flashinfer/actions/runs/23589405658/job/68979382814?pr=2756

This blocks some other test cases. You can run pre-commit this way: https://github.com/flashinfer-ai/flashinfer/blob/31b63bc3d73597612419b4831218feabc57fd97d/CLAUDE.md?plain=1#L20

Please let me know if you meet any question~
